### ISSUE TITLE
Add the ability to re-detect all face for all image

### ISSUE THREAD

**Issue by vsakhnenko** [photoview/photoview#1071]
As far as I know, detection only occurs during image processing, it would be nice to be able to do it whenever you want.
What is needed:

- clear face_group table
- clear image_faces table
- re-detect all image using already implemented functionality
- add button to front-end (i think it would be find add it near SCAN in settings)

This can help with the error in case of failed detection (see #1063)

**Comment by kkovaletp** [comment]
@vsakhnenko, thanks for the proposal.
In general, the ability to re-detect faces on already scanned photos is a good idea, but I see the next concerns in the proposal:
- clearing tables means losing all user data. I'd avoid this in the product. If needed, an admin can do this directly in the DB. As Photoview is a read-only media gallery, the only sensitive data we have, is related to faces: users have to name and merge them manually, which takes a lot of time. If this button is pressed by mistake or because of a misunderstanding of all the consequences, the user data is lost
- for the case of adding such a button to the photo details pane, it looks less destructive, but still: what will happen with already triaged faces? It seems to be a valid improvement, but we need to design the solution taking into account all the corner cases

I'd say that clearing and re-detecting unlabeled faces and groups looks not so destructive, as no manual efforts are expected to be spent on them. What do you think?

**Comment by vsakhnenko** [comment]
> I'd say that clearing and re-detecting unlabeled faces and groups looks not so destructive, as no manual efforts are expected to be spent on them. 
 I don't understand this sentence, but I agree that the idea of doing re-detection on all images is not so good (if someone has a lot of photos and presses the button by mistake, it will be terrible).
 In case of repeated detection in photo (I may be wrong, but I think this will be enough):

-  for photo where was re-detection run delete (or implement something like save delete by adding new column to db) data only from image_face by media_id. Also check if in image_face it was last record with such face_group_id delete face_group as well (we dont need face_group if it is not connect to any photo).
- if no faces have been previously detected in the photo, do not allow re-detection.
- consider allowing re-detection only for users with admin rights.
- when user press button for re-detection warn about the possibility of data loss that cannot be recovered later(it may be worth considering adding a pop-up window with a field for re-entering the password to ensure that it is accurately performed by the admin)
what do you think about it? If you think it would be useful, I can start working on it

**Comment by kkovaletp** [comment]
> I don't understand this sentence

I mean that it is safer to delete unlabeled face groups and their faces, as no manual work has been done there.

> * if no faces have been previously detected in the photo, do not allow re-detection.

Why? There is nothing to delete, but why not try to detect faces in the photo? Maybe the previous time it wasn't detected by mistake/bug...

> * consider allowing re-detection only for users with admin rights.

This won't work, as Photoview allows different media libs for different users. In this case, only non-admin users might have access to the particular media.

> * when user press button for re-detection warn about the possibility of data loss that cannot be recovered later(it may be worth considering adding a pop-up window with a field for re-entering the password to ensure that it is accurately performed by the admin)

This is a good idea. I'd show the list of items, which are going to be removed (with additional warning if a labeled face group is in the list) so that users can give informed confirmation. Also, I'd ask for the user password, as this is a destructive action.

> * what do you think about it? If you think it would be useful, I can start working on it

Sure. I guess that this feature might be useful for users. 
@googollee, please review our discussion and share your thoughts, so that the implementation is going to be done in the right way.

**Comment by googollee** [comment]
This issue describes a solution, but not a problem. How to trigger re-detecting? Should we add a button in UI? In which page?

#269 is a real problem, and https://github.com/photoview/photoview/issues/269#issuecomment-807131290 is a better solution.

#1063 is a bug in the code, and re-detecting doesn't fix that bug, just hides it.

**Comment by vsakhnenko** [comment]
In my opinion, it will not hide the bug, but will allow us to fix it at the moment when we do not have enough information about it. I plan to add more logs. Plus, we will add a new feature to redetect faces (if it was detected incorrectly for someone or was not detected at all for some reason).

But on my part, I just want to be useful and try to do something for this project, if you have a better idea how to do it, write and I try to implement it. Thanks for you time

**Comment by vsakhnenko** [comment]
We agreed to redaction for only one photo above 

> > I'd say that clearing and re-detecting unlabeled faces and groups looks not so destructive, as no manual efforts are expected to be spent on them.
> 
> I don't understand this sentence, but I agree that the idea of doing re-detection on all images is not so good (if someone has a lot of photos and presses the button by mistake, it will be terrible). In case of repeated detection in photo (I may be wrong, but I think this will be enough):
> 
> * for photo where was re-detection run delete (or implement something like save delete by adding new column to db) data only from image_face by media_id. Also check if in image_face it was last record with such face_group_id delete face_group as well (we dont need face_group if it is not connect to any photo).
> * if no faces have been previously detected in the photo, do not allow re-detection.
> * consider allowing re-detection only for users with admin rights.
> * when user press button for re-detection warn about the possibility of data loss that cannot be recovered later(it may be worth considering adding a pop-up window with a field for re-entering the password to ensure that it is accurately performed by the admin)
>   what do you think about it? If you think it would be useful, I can start working on it

**Comment by vsakhnenko** [comment]
But if you think it's not necessary for the product, is there something else where I can be useful?

**Comment by googollee** [comment]
I agree on https://github.com/photoview/photoview/issues/1071#issuecomment-2391261195. Could you also update the title and the main description? It's easy to find this conclusion.

**Comment by vsakhnenko** [comment]
I can start with the implementation and then we can decide if it is good enough

**Comment by kkovaletp** [comment]
> This issue describes a solution, but not a problem. How to trigger re-detecting? Should we add a button in UI? In which page?
> 
> #269 is a real problem, and [#269 (comment)](https://github.com/photoview/photoview/issues/269#issuecomment-807131290) is a better solution.
> 
> #1063 is a bug in the code, and re-detecting doesn't fix that bug, just hides it.

I agree with you, but:
- to me it looks like these 2 issues complement each other: this one is about an improvement of existing functionality, while the #269 is a bug in dealing with garbage, falsely reported by the face detector.
- How to trigger re-detecting? We're trying to discuss the possible implementations and their pros/cons/limitations. For now, I like the idea of the button being on the media file details pane - to trigger re-detection for the particular photo. Also, we "safely" can run it for all lib while applying to unlabeled groups only, as no manual work has been done for them. What to do with the labeled groups and their faces - is an open question.
- This feature won't actually hide the #1063, it might revert the state to let the user try again in some cases, but actually, the bug can be still reproduced with this change.
- regarding the solution, proposed in the #269: we need to think about the next side effect: if there is a hidden group and a new face is detected, there is a chance to merge it automatically with that group, so as a result, a person won't be visible, even while the face was detected. When we go to the implementation of the #269 solution, we have to design the functionality to prevent or mitigate this case.

**Comment by vsakhnenko** [comment]
Hi, I have implemented the basic logic and recorded some videos of how it works, can you check when you have time?

1. Delete all faces_image and face_group and re-detect
2. Set a label for the face_group and after re-detecting it on another photo, it should join the existing face_group
3. I will simulate an error that some users have encountered
[Uploading 3.webm…
[2.webm](https://github.com/user-attachments/assets/f9c12496-6c19-4165-9408-903ae2345fd1)
[1.webm](https://github.com/user-attachments/assets/de920dca-b7c1-459a-8cf5-f81fd7d073b5)
]()

Work in progress:
1. Ask the user for a password before re-detection
2. Improve the look of the front end, I'm not very good at design and if you have a better idea how it should look, write
3. Write down  scenarios I need to test, (mb i don't know about some of them)

**Comment by vsakhnenko** [comment]
Now everything is implemented, I'm waiting for idea for improvements  the design and mb some scenarios which i can test. (All what i can imagine i tested)

**Comment by kkovaletp** [comment]
I'll check them in detail in a few days, as I'm pretty busy now. Sorry for the delay. Just want you to know that I saw your comments, but didn't have time to analyze them yet

**Comment by kkovaletp** [comment]
@vsakhnenko, I finally got to this ticket, and here are my comments:

> 1\. Delete all faces_image and face_group and re-detect

- The re-detect button should be located in the "People" section of the photo's info under all the paces if any; be named, like "Re-detect unlabeled faces"; look like other buttons there ("add shares", "set as album cover"), but have orange text color to indicate potentially destructive changes.
- Once pressed, it would be good to see a pop-up outside of the sidebar with the list of all face groups, affected by this (with their face circles) and a warning message that faces, detected on this photo and being part of these face groups, are going to be removed and re-detected on the photo. Then the newly-detected faces will be auto-merged to the best matching groups, which might not be the current ones.
- On the 1st video you delete all the faces and groups irrespectively to the ones, related to the current photo. This is not what we've discussed before. I'd expect, only faces from the current photo, being part of unlabeled groups, to be deleted. Then, only the groups with no faces left to be deleted as well. That's why you had to re-detect faces on the 2nd photo even while this is the same person. In this particular case, I'd expect that the face from 2nd photo and this group should not be deleted, even if it is unlabeled.

> 2\. Set a label for the face_group and after re-detecting it on another photo, it should join the existing face_group

- You didn't show the result in the DB after the re-detection, however, you did not do what is described in the 2nd point: you labeled 1st and 2nd face groups, removed 3rd one with its group, re-detected faces, which actually added the 3rd face with the group. No changes were made (and expected) to the 2nd photo, as this re-added face is not part of it.
- To execute the scenario, you had to manually delete 1 of the faces for the 1st group, then re-detect them and check that the manually removed face was auto-merged into the 1st group

> 3\. I will simulate an error that some users have encountered

As far as I see, you didn't get the error. I was able to reproduce it from the People page. Not from a photo's info.

> 1\. Ask the user for a password before re-detection

As it is still not completed, there might be the case, it is not yet implemented, but:
- if the password is incorrect, show a validation error under the password field. If it is for the 3rd time, replace the content of the pop-up with an error message that the password is incorrect 3 times and the user is logged out; show the "Login" button under the message; logout the user on the background in a similar way to the logout button in the Settings, but don't refresh the page and don't redirect to the login screen,  so that users can see and read the error; once the login button is pressed, refresh the page and (as there is no active session) the login page has to be shown.

> 2\. Improve the look of the front end, I'm not very good at design and if you have a better idea how it should look, write

I'm not a designer as well) just try to follow the product design when adding or changing something in the UI. I'll provide the comments and suggestions if I'll see something not looking good for me) Anyway, feel free to discuss UI here.

> 3\. Write down  scenarios I need to test, (mb i don't know about some of them)

Well, it would be great to test:
- all touched code files - with unit/integration tests
- all previously discussed and possible from the UI and code perspective use cases - manually

**Comment by kkovaletp** [comment]
> Now everything is implemented

You could create a PR, linked to this issue, so we could see your implementation and better understand its logic.
Set it to Draft state until it is ready for review.

**Comment by vsakhnenko** [comment]
@kkovaletp okay i will try to fix what you write and create PR

**Comment by vsakhnenko** [comment]
Hi do you have any formatter for frontend code?

**Comment by kkovaletp** [comment]
> formatter for frontend code

No, we have only ESlint integrated. I'd suggest using IDE's formatted functionality (at least there is one in VScode and various Intellij IDEs)

**Comment by vsakhnenko** [comment]
#1101
